app/frontend/wallet/account.ts

```diff
@@ -294,30 +295,43 @@ const Account = ({
     return _getMaxSendableAmount(utxos, address, sendAmount)
   }
 
-  const selectUTxOs = async (txType: TxType): Promise<UTxO[]> => {
-    const availableUtxos = await getUtxos()
-    const nonStakingUtxos = availableUtxos.filter(({address}) => !isBase(addressToHex(address)))
-    const baseAddressUtxos = availableUtxos.filter(({address}) => isBase(addressToHex(address)))
+  const selectUTxOs = (utxos: UTxO[], txPlanArgs: TxPlanArgs): UTxO[] => {
+    const nonStakingUtxos = utxos.filter(({address}) => !isBase(addressToHex(address)))
+    const baseAddressUtxos = utxos.filter(({address}) => isBase(addressToHex(address)))
+    const adaOnlyUtxos = baseAddressUtxos.filter(({tokenBundle}) => tokenBundle.length === 0)
+    const tokenUtxos = baseAddressUtxos.filter(({tokenBundle}) => tokenBundle.length > 0)
     const randomGenerator = PseudoRandom(seeds.randomInputSeed)
-    // TODO: ideally we want to do nonStaking, with token, only ada, the rest
-    if (txType === TxType.SEND_ADA || txType === TxType.CONVERT_LEGACY) {
+
+    if (
+      txPlanArgs.txType === TxType.SEND_ADA &&
+      txPlanArgs.sendAmount.assetFamily === AssetFamily.TOKEN
+    ) {
+      const {policyId, assetName} = txPlanArgs.sendAmount.token
+      const utxosWithToken = tokenUtxos.filter(({tokenBundle}) =>
+        tokenBundle.some((token) => token.policyId === policyId && token.assetName === assetName)
+      )
+      const utxosWithoutToken = tokenUtxos.filter(
+        ({tokenBundle}) =>
+          !tokenBundle.some((token) => token.policyId === policyId && token.assetName === assetName)
+      )
       return [
+        ...shuffleArray(utxosWithToken, randomGenerator),
```

Comment:
general note - shuffling the individual "subcategories" of tokens has no benefit for the purpose of utxo "self-organization" as opposed to shuffling all the tokens in the beginning and then filtering the subcategories (provided the filtering respects the original order of elements which it does)

Reply:
Hm, you are right, given that the filtering is respective to the shuffling, it should be enough to shuffle them all in the beginning. It didn't break any of the tests so I remove the shuffling of the filtered utxos.